src/instructlab/sdg/sdg.py

```diff
@@ -1,21 +1,88 @@
 # SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import Optional
+
 # Third Party
-from datasets import Dataset
+from datasets import Dataset, concatenate_datasets
 
 # Local
 from .pipeline import Pipeline
 
 
 # This is part of the public API.
 class SDG:
-    def __init__(self, pipelines: list[Pipeline]) -> None:
+    def __init__(
+        self,
+        pipelines: list[Pipeline],
+        num_workers: Optional[int] = 1,
+        batch_size: Optional[int] = None,
+    ) -> None:
+        """
+        Initialize a new instance of SDG with a collection of pipelines to run
+        in sequence.
+
+        pipelines (list[Pipeline]): A list of pipeline objects to run in
+            sequence.
+        num_workers (int | None): The number of workers to use. None implies the
+            default behavior of min(32, os.cpu_count() + 4)
+            https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+        batch_size (int | None): The batch size to use when processing. None
+            implies no batching.
+        """
         self.pipelines = pipelines
+        self.num_workers = num_workers
+        self.batch_size = batch_size
 
     def generate(self, dataset: Dataset):
         """
         Generate the dataset by running the chained pipeline steps.
-        dataset: the input dataset
+        dataset (Dataset): the input dataset
+        """
+        input_splits = self._split_dataset(dataset)
+        output_splits = [None] * len(input_splits)
+
+        with ThreadPoolExecutor(max_workers=self.num_workers) as executor:
```

Comment:
This has me wondering:

```
All threads enqueued to ThreadPoolExecutor will be joined before the interpreter can exit. Note that the exit handler which does this is executed before any exit handlers added using atexit. This means exceptions in the main thread must be caught and handled in order to signal threads to exit gracefully. For this reason, it is recommended that ThreadPoolExecutor not be used for long-running tasks.
```

Could you experiment with killing the parent while the pipeline is running, to check the behavior is graceful?

On the other hand, if we used ProcessPoolExecutor, I suspect we might run into pickling issues like I fixed in commit 7cfbaa9

Reply:
Yeah, `ProcessPoolExecutor` would very likely cause pickling problems. I thought about making it configurable, but decided against it to keep it simple. Probably worth a simple experiment to see if a `Dataset` can easily pass between processes (my money is on "no").

Reply:
Ok, looks like `Dataset` itself is actually pretty happy to pass between processes:

*tmp.py*

```py
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
from datasets import Dataset
import os

ds = Dataset.from_list([{"foo": 1}, {"foo": 2}])

pool_type = ThreadPoolExecutor if int(os.getenv("THREADS", "0")) else ProcessPoolExecutor

def doit(dataset):
    return dataset.map(lambda r: {"foo": r["foo"] * 2})

if __name__ == "__main__":
    print(f"Pool Type: {pool_type}")
    with pool_type() as pool:
        fut = pool.submit(doit, ds)
        print(fut.result().to_list())
```

```sh
?> python tmp.py
Pool Type: <class 'concurrent.futures.process.ProcessPoolExecutor'>
Map: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 2649.59 examples/s]
[{'foo': 2}, {'foo': 4}]

?> THREADS=1 python tmp.py
Pool Type: <class 'concurrent.futures.thread.ThreadPoolExecutor'>
Map: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1261.26 examples/s]
[{'foo': 2}, {'foo': 4}]
```

Reply:
I do suspect that if we put any non-primitives into the `Dataset` (not sure if this is even allowed based on using `pyarrow` behind the scenes), we'd hit pickling problems like you did with `SSLContext`.

Reply:
>Could you experiment with killing the parent while the pipeline is running, to check the behavior is graceful?

I was thinking about this a little. I'll see what I can do to drum up a full test of this. I suspect the behavior would be to block termination (short of `SIGKILL`) until all non-error threads have fully terminated.

In `caikit`, we built a wrapper ([DestroyableThread](https://github.com/caikit/caikit/blob/main/caikit/core/toolkit/concurrency/destroyable_thread.py#L42)) to handle this so that a given piece of work running in a thread could be cancelled without destroying the thread itself from the pool. This is probably overkill since `caikit` is designed to be a long-running server and a single dead task should not cause harm to the central thread pool. In the case of SDG, it could still be useful to signal that other `future` shards should stop early if one of the shards has failed, though I suspect that a simple `SIGKILL` to the parent process would do the trick since this will all be driven by the CLI.

Reply:
Ok, as suspected, it looks like `SIGTERM` will wait for a "graceful" exit which means finishing all running futures, while `SIGKILL` will unceremoniously kill everything:

<details>
<summary>tmp.py</summary>

```py
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor, as_completed
from datasets import Dataset
import os
import time
import atexit
from functools import partial

pool_type = ThreadPoolExecutor if int(os.getenv("THREADS", "0")) else ProcessPoolExecutor

def doit(dataset, sleep_time=0, should_raise=False):
    if sleep_time:
        time.sleep(sleep_time)
    if should_raise:
        raise RuntimeError("Yikes")
    return dataset.map(lambda r: {"foo": r["foo"] * 2})

def end_report(futures):
    print("Done: " + str([f"{i}: {fut.done()}" for i, fut in enumerate(futures)]))
    print("Cancelled: " + str([f"{i}: {fut.cancelled()}" for i, fut in enumerate(futures)]))
    print("Exception: " + str([f"{i}: {fut.exception(timeout=0.01)}" for i, fut in enumerate(futures)]))

if __name__ == "__main__":
    print(f"Pool Type: {pool_type}")
    ds1 = Dataset.from_list([{"foo": 1}, {"foo": 2}])
    ds2 = Dataset.from_list([{"foo": 3}, {"foo": 4}])
    ds3 = Dataset.from_list([{"foo": 5}, {"foo": 6}])
    with pool_type() as pool:
        futures = [
            pool.submit(doit, ds, sleep_time=i * 3, should_raise=not i)
            for i, ds in enumerate([ds1, ds2, ds3])
        ]
        atexit.register(partial(end_report, futures))
        for future in as_completed(futures):
            print(future.result().to_list())
```
</details>

**With SIGTERM**

```sh
time THREADS=1 python tmp.py & proc=$! && kill $proc
[1] 24619
Pool Type: <class 'concurrent.futures.thread.ThreadPoolExecutor'>
Map: 100%|████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 857.29 examples/s]
Map: 100%|████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 818.96 examples/s]
Traceback (most recent call last):
  File "/Users/ghart/Projects/github/instructlab/sdg/tmp.py", line 34, in <module>
    print(future.result().to_list())
          ^^^^^^^^^^^^^^^
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ghart/Projects/github/instructlab/sdg/tmp.py", line 14, in doit
    raise RuntimeError("Yikes")
RuntimeError: Yikes
Done: ['0: True', '1: True', '2: True']
Cancelled: ['0: False', '1: False', '2: False']
Exception: ['0: Yikes', '1: None', '2: None']

real	0m6.612s
user	0m0.741s
sys	0m3.233s

[1]+  Exit 1                  time THREADS=1 python tmp.py
```

**With SIGKILL**
(NOTE: This _also_ kills `time`, but it all happens "instantly")
```sh
?> time THREADS=1 python tmp.py & proc=$! && kill -9 $proc
[1] 24510
[1]+  Killed: 9               time THREADS=1 python tmp.py
```

Reply:
Also, letting it all run, including the exception raised in the first future, waits for the others to complete (as it should):

```sh
time THREADS=1 python tmp.py 
Pool Type: <class 'concurrent.futures.thread.ThreadPoolExecutor'>
Map: 100%|███████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1006.07 examples/s]
Map: 100%|████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 842.74 examples/s]
Traceback (most recent call last):
  File "/Users/ghart/Projects/github/instructlab/sdg/tmp.py", line 34, in <module>
    print(future.result().to_list())
          ^^^^^^^^^^^^^^^
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/ghart/mambaforge/envs/instructlab-sdg/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ghart/Projects/github/instructlab/sdg/tmp.py", line 14, in doit
    raise RuntimeError("Yikes")
RuntimeError: Yikes
Done: ['0: True', '1: True', '2: True']
Cancelled: ['0: False', '1: False', '2: False']
Exception: ['0: Yikes', '1: None', '2: None']

real	0m6.584s
user	0m0.672s
sys	0m3.332s
```

Reply:
> Ok, as suspected, it looks like `SIGTERM` will wait for a "graceful" exit which means finishing all running futures

Ok, thanks. I guess that works as a reasonable UX for the CLI